File: LeetCode/Linked-List-Cycle.cpp
// Given a linked list, determine if it has a cycle in it.

#include <iostream>
using namespace std;

struct ListNode {
  int val;
  ListNode *next;
  ListNode(int x) : val(x), next(NULL) {}
};

bool hasCycle(ListNode * head) {
  if(head == NULL) return false;

  ListNode *slow = head;
  ListNode *fast = head;

  while(fast->next && fast->next->next) {
    slow = slow->next;
    fast = fast->next->next;
    if(slow == fast) { return true; }
    return false;
  }
}

File: Python-Programming/Fast-and-Slow-Pointers/Linked-List-Cycle.py
class Node:
    def __init__(self, value, next=None):
        self.value = value
        self.next = next

    def has_cycle(head):
        slow, fast = head, head
        while fast is not None and fast.next is not None:
            fast = fast.next.next
            slow = slow.next
            if slow == fast:
                return True  # found the cycle

        return False
